### ISSUE TITLE
Discussion: CI/CD

### ISSUE THREAD

**Issue by kartik-budhiraja** [Seneca-CDOT/telescope#6]
The MVP needs to have a CI/CD pipeline set up with automatic tests and deployments.
We need to decide between the following options:
- Github Actions
- Travis CI
- Circle CI

I suggest going for github actions because other options would require setting up external accounts while for github actions, its all here on github only.
On the other side, github actions is still in beta but fairly stable stage.

**Comment by humphd** [comment]
Because we have so many people who want to be involved, and everyone is learning technologies as they go, we could also consider implementing multiple pipelines.  Obviously we can't do deploy multiple times, but linting, unit tests, performance testing, fuzz testing, etc. could be done in parallel on all of GH Actions, Travis, Circle CI, etc.

Later on we can finalize on one if we want, but having extra testing will only help in the early stages.

**Comment by UltimaBGD** [comment]
That may lead to a bit more maintenance with multiple pipelines, but with a larger group like this would probably okay, especially if people specialize in maintenance of 1 or 2 pipelines. May not be the most efficient way, but probably far more educational.

**Comment by c3ho** [comment]
I recall the prof mentioning he doesn't expect us to complete this project, but it should be more of a learning experience/experimenting. I'm all for learning different pipelines.

**Comment by ragnarokatz** [comment]
I've had experience with [jenkins pipeline setup w/docker](https://hub.docker.com/r/jenkinsci/blueocean/) before but haven't played around with the others as much. I'm particular interested in **travis CI** as well as using **sonarqube** for code scans.

**Comment by humphd** [comment]
Filed #32, which we can use as our first CI automation target.  Once we get that landed, we can spin up various CI platforms to automatically build every PR and merge commit.  Later we can expand this to full tests.

**Comment by kartik-budhiraja** [comment]
I will spin up the pipeline on Travis CI and GitHub actions as soon as the eslint config is added!

**Comment by humphd** [comment]
@kartik-budhiraja would you be willing to file separate issues for each of Travis, GH Actions, CircleCI, and any others people want to try?  Ideally we'll let them get done by individuals vs. one person doing multiple (i.e., pick one of travis or gh, but not both).

Thanks for offering to work on this, I'd love to get the CI pipeline flowing.

**Comment by kartik-budhiraja** [comment]
@humphd  Sure, I filed three separate issues (#43, #44, #45) and I will be taking over only travis setup.

**Comment by dbeigi** [comment]
@humphd is it possible to help someone with their pipeline?

**Comment by wajeehsheikh** [comment]
Hi all,

Seeing that David is open to multiple CI
I would like to work on [Jenkins](https://jenkins.io)

Thank you,
Waj

**Comment by wajeehsheikh** [comment]
> @humphd is it possible to help someone with their pipeline?

HI @dbeigi,
You can help me with setting up Jenkins for CI/CD if you want.
#98 Set up CI/CD pipeline at Jenkins.

Thank you,
Waj

**Comment by dbeigi** [comment]
@wajeehsheikh 👍

**Comment by humphd** [comment]
We've got good tickets covering the specifics of this now, closing in favour of those.
